### PR TITLE
[Federation] Hpa controller controls target objects

### PR DESCRIPTION
**Pull request by irfanurrehman**
This is in the series of PRs over https://github.com/kubernetes/kubernetes/pull/45993.
The last commit is reviewable. Probably the last PR in this chain with e2e tests for relevant scenario, including the scenario created by this PR is soon to follow.

**Special notes for your reviewer**:
@kubernetes/sig-federation-pr-reviews 
@quinton-hoole 

**Release note**:

```NONE
```

### CODE REVIEW

File: federation/pkg/federatedtypes/hpa.go
Comment (quinton-hoole) [inline federation/pkg/federatedtypes/hpa.go]: This and the above function seem rather brittle?  It seems that you should use the list of object types supported by the cluster HPA's ideally (i.e. not hardcode the names of the types here, and have to update them here every time the cluster HPA's get updated.  I realize that this is perhaps non-trivial, and am happy to defer it to a later PR provided that we open a known issue for it now.

File: federation/pkg/federatedtypes/scheduling.go
Comment (quinton-hoole) [inline federation/pkg/federatedtypes/scheduling.go]: Rather replace with a simple set inclusion operation?  That would be less code, and more efficient.  I think we have a function in the utils library to do that.  StringSet or similar.

**Review by ghost — CHANGES_REQUESTED** [review]
More review to come...

File: federation/pkg/federatedtypes/hpa.go
Comment (irfanurrehman) [inline federation/pkg/federatedtypes/hpa.go]: Actually, I did look to do something similar what you suggest, but to my surprise found that, at almost all places, where ever the kind or the resource name is initialised, its used as a string directly.
For example [federation-apiserver default resource config](https://github.com/kubernetes/kubernetes/blob/master/federation/cmd/federation-apiserver/app/server.go#L267) (which is done quite similar in kube-apiserver) or specifying a kind, [as in deployment controller ref kind](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/deployment_controller.go#L417) or [the replicaset controller kind itself](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replicaset/replica_set.go#L62).
I did go ahead to use the string as in the code reviewed, thus.

Additionally, in k8s api, the hpa targetref object is not interpreted (as is done here) within the hpa controllers purview. The scales subresource is used instead. So what ever target name and kind comes up in the hpa spec, the scales api is called using the same. Only particular types support this subresource, thus the api call will fail, if called on a wrong type or kind (which probably was wrong in the hpa spec, specified by the user), [ref](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podautoscaler/horizontal.go#L384). 
I did use only "Deployment" and "ReplicaSet" in the kind check here, because those are the only two kinds which are available against federation API.
I am not sure, how else can I handle it. On the other hand, I have limited understanding of the API machinery; @nikhiljindal would you kindly be able to help a bit on this, thanks!

File: federation/pkg/federatedtypes/scheduling.go
Comment (irfanurrehman) [inline federation/pkg/federatedtypes/scheduling.go]: If you are talking about [this](https://github.com/kubernetes/apimachinery/blob/master/pkg/util/sets/string.go#L19), I found it less efficient then using the simple slice like I have used here, simply by the fact that it uses maps (I have used it in hpa controller though, because of the need of set operations over there).
I could have used this string set to store the cluster list in the first place to ensure, I don't introduce a function like the one, I got this comment on. But a slice suited the need better because of marshalling being super easy [as here](https://github.com/kubernetes/kubernetes/pull/49950/commits/41b6440a938c00b3b8b40881bed986b1304cdb71#diff-153c05d91ab301e028511701ea78219bR39).

File: federation/pkg/federation-controller/util/hpa/hpa.go
Comment (quinton-hoole) [inline federation/pkg/federation-controller/util/hpa/hpa.go]: OK, so rather than overwrite the user-specified cluster list on the target object, you add another list that overrides the former, right?

**Review by ghost — CHANGES_REQUESTED** [review]
First pass review complete.  You seem to be missing unit tests on some of this code?

File: federation/pkg/federatedtypes/hpa.go
Comment (irfanurrehman) [inline federation/pkg/federatedtypes/hpa.go]: I omitted that because, I was already formatting an error string in the functions `getRuntimeObjectForKind()` and `updateRuntimeObjectForKind()`. On second thoughts, and your suggestion, have augmented the string here also.

File: federation/pkg/federatedtypes/hpa.go
Comment (irfanurrehman) [inline federation/pkg/federatedtypes/hpa.go]: Updated!

File: federation/pkg/federatedtypes/hpa.go
Comment (irfanurrehman) [inline federation/pkg/federatedtypes/hpa.go]: Updated!

File: federation/pkg/federatedtypes/hpa.go
Comment (irfanurrehman) [inline federation/pkg/federatedtypes/hpa.go]: Yes, I completely agree, but as mentioned I did not find a better way for this.
Also, I have used Deployment and ReplicaSet only in the target objects, because those are the two objects which are supported in control plane as of now. I have added comments on top of the functions for now, with a TODO to update the documentation when I do that.

File: federation/pkg/federatedtypes/scheduling.go
Comment (irfanurrehman) [inline federation/pkg/federatedtypes/scheduling.go]: Updated ReplicaSchedulingStatus to ReplicaStatus.
ReplicaScheduleState is the result of asapter specific schedule() (eg planner.plan for replicasets) function, which is then used to update objects into clusters.
ReplicaStatus contains the details of status fields from the cluster objects, which need accumulation to update the status of the federated object. 
Updated the comments for the same too.

File: federation/pkg/federatedtypes/scheduling.go
Comment (irfanurrehman) [inline federation/pkg/federatedtypes/scheduling.go]: Ack.

File: federation/pkg/federatedtypes/scheduling.go
Comment (irfanurrehman) [inline federation/pkg/federatedtypes/scheduling.go]: Updated!

File: federation/pkg/federatedtypes/scheduling.go
Comment (irfanurrehman) [inline federation/pkg/federatedtypes/scheduling.go]: I completely get your suspicion.
I had 2 options:
1. Use the cluster selector annotation itself to select the clusters to which the hpa target object should go (decided by hpa controller), overriding the user specified cluster selector (if any). In effect the hpa objects cluster selector overrides and applies to the target object also and hpa controllers decisions also apply, subsequently. This can be confusing to the user (who specified the original selector annotations on the target object) see it automagically change.
2. Let the cluster selectors on both objects work independently and use a separate annotation to communicate the hpa's intent. There can be a situation that both the selectors do not intersect, in which case no cluster will have a target object (but local hpas will exist). The fed hpa's decisions are made on observations of its local hpas, which originally get created as per the hpa cluster selector, and if the two selectors intersect, the hpa replicas would get concentrated into the intersection of clusters (which also would have target objects and thus actual status fields; currentReplicas, desiredReplicas, etc. would be updated). This result can also be confusing to the user, but does not override his/her choices. Ideally the user who creates the hpa, also created the target object and ideally should put the same cluster selector on both, but if he/she doesn't do that, the results are predictable.
I have chosen the latter (which appeared sound to me when I coded it :) ), with an intent that this is updated in docs explaining the consequences.
If you think there probably is a better way, we can have a short chat to discuss the same.

File: federation/pkg/federatedtypes/scheduling.go
Comment (irfanurrehman) [inline federation/pkg/federatedtypes/scheduling.go]: Please see my comment above.

File: federation/pkg/federatedtypes/scheduling.go
Comment (irfanurrehman) [inline federation/pkg/federatedtypes/scheduling.go]: ReplicaScheduleState stores the result of schedule() function specific to adapter (which cluster gets how many replicas). 0 means the particular cluster does not get any replicas, and is the initialised state too. Thought better to use 0  as that marker to avoid null checks. I have used an additional bool field to indicate if this cluster is selected in this reconcile pass or not, because we use the same field(s) to map the selection percolated from hpa or actual schedule function of this adapter.

File: federation/pkg/federatedtypes/scheduling.go
Comment (irfanurrehman) [inline federation/pkg/federatedtypes/scheduling.go]: I think this was confusing because of the naming of the two structures also. I have updated the name with comments explaining the difference to avoid confusion [also explained in this comment](https://github.com/kubernetes/kubernetes/pull/49950/files#r1332031210).

File: federation/pkg/federatedtypes/scheduling.go
Comment (irfanurrehman) [inline federation/pkg/federatedtypes/scheduling.go]: As above.

File: federation/pkg/federation-controller/util/hpa/hpa.go
Comment (irfanurrehman) [inline federation/pkg/federation-controller/util/hpa/hpa.go]: It does not override the former but the intersection of the two is used as elaborated [here](https://github.com/kubernetes/kubernetes/pull/49950/files#r133218666).

**Comment by irfanurrehman** [comment]
@quinton-hoole updated this based on your comments PTAL!

File: federation/pkg/federatedtypes/hpa.go
Comment (irfanurrehman) [inline federation/pkg/federatedtypes/hpa.go]: This is actually used only for crud test written with the sync controller (which tests crud of the hpa object alone). Meanwhile, I will add tests for both types in e2e.

File: federation/pkg/federatedtypes/scheduling.go
Comment (quinton-hoole) [reply federation/pkg/federatedtypes/scheduling.go]: Yes, what you have chosen seems sensible to me.

File: federation/pkg/federatedtypes/scheduling.go
Comment (quinton-hoole) [reply federation/pkg/federatedtypes/scheduling.go]: I think you've done the right thing (option 2).

File: federation/pkg/federatedtypes/scheduling.go
Comment (quinton-hoole) [reply federation/pkg/federatedtypes/scheduling.go]: OK, that makes sense.

File: federation/pkg/federatedtypes/scheduling.go
Comment (quinton-hoole) [reply federation/pkg/federatedtypes/scheduling.go]: OK, I think that's fine.  If Kubernetes represents unknown status as zero, we should probably follow that for now.

File: federation/pkg/federation-controller/util/hpa/hpa.go
Comment (quinton-hoole) [reply federation/pkg/federation-controller/util/hpa/hpa.go]: OK, thanks, makes sense.

**Review by ghost — APPROVED** [review]
Thanks @irfanurrehman .  The code looks good to me.  I would like someone else to also do a code review before I approve for merging.  I'll solicit help on slack.

**Comment by quinton-hoole** [comment]
No help offered on slack.  I've added it to the SIG meeting notes as a PR requiring some additional love.

**Comment by irfanurrehman** [comment]
@quinton-hoole this needs lgtm as well, thanks a lot for handling this!

**Comment by jdumars** [comment]
@quinton-hoole @irfanurrehman can you follow the exception process to get this in 1.8?  https://github.com/kubernetes/features/blob/master/EXCEPTIONS.md Also, please make sure this feature is mentioned in the https://github.com/kubernetes/features/blob/master/release-1.8/release_notes_draft.md Thanks!

**Comment by irfanurrehman** [comment]
> @quinton-hoole @irfanurrehman can you follow the exception process to get this in 1.8? https://github.com/kubernetes/features/blob/master/EXCEPTIONS.md Also, please make sure this feature is mentioned in the https://github.com/kubernetes/features/blob/master/release-1.8/release_notes_draft.md Thanks!

@jdumars I did send in an exception request yesterday; please let me know if anything else might be needed. thanks!

**Comment by jdumars** [comment]
@quinton-hoole if you (and SIG Federation) LGTM/Approve this, it will go in 1.8 assuming you get it done today.  Tomorrow is likely the hard cutoff for 1.8 except extraordinary fixes that need to be cherry picked in.

**Pull request merged**
